libvast/builtins/operators/summarize.cpp

```diff
@@ -321,100 +209,282 @@ struct group_by_key_equal {
-        new_bucket->push_back(std::move(*function));
+      new_bucket->group_by_types.reserve(bound.group_by_columns.size());
+      for (auto&& column : bound.group_by_columns) {
+        if (column) {
+          new_bucket->group_by_types.push_back(
+            group_type::make_active(column->type));
+        } else {
+          new_bucket->group_by_types.push_back(group_type::make_empty());
+        }
+      }
+      new_bucket->aggregations.reserve(bound.aggregation_columns.size());
+      for (auto col = size_t{0}; col < bound.aggregation_columns.size();
+           ++col) {
+        // If this aggregation column exists, we create an instance of the
+        // aggregation function with the type of the column. If it does not
+        // exist, we store `std::nullopt` instead of an aggregation function, as
+        // we will later use this as a signal to set the result column to null.
+        if (bound.aggregation_columns[col].has_value()) {
+          auto input_type = bound.aggregation_columns[col]->type;
+          if (auto instance
+              = config.aggregations[col].function->make_aggregation_function(
+                input_type)) {
+            new_bucket->aggregations.push_back(
+              aggregation::make_active(std::move(*instance)));
+          } else {
+            // We already emitted a warning for this earlier.
+            new_bucket->aggregations.push_back(aggregation::make_dead());
+          }
+        } else {
+          // If the column does not exist, we cannot instantiate the function
+          // yet because we don't know which type to use.
+          new_bucket->aggregations.emplace_back(aggregation::make_empty());
+        }
       }
       auto [it, inserted] = buckets.emplace(materialize(reusable_key_view),
                                             std::move(new_bucket));
       VAST_ASSERT(inserted);
       return it.value().get();
     };
-    // Iterate over all rows of the batch, and determine a sliding window of
-    // rows beloging to the same batch as large as possible, and then update the
-    // corresponding bucket.
-    auto first_row = 0;
+    // This lambda is called for consecutive rows that belong to the same group
+    // and updates its aggregation functions.
+    auto update_bucket = [&](bucket& bucket, int64_t offset, int64_t length) {
+      for (auto [aggr, input] :
+           zip_equal(bucket.aggregations, aggregation_arrays)) {
+        if (!input) {
+          // If the input column does not exist, we have nothing to do.
+          continue;
+        }
+        if (!aggr.is_active()) {
+          // If the aggregation is dead, we have nothing to do. If it is
+          // empty, we know that the aggregation column does not exist in
+          // this schema, and thus have nothing to do as well. The only
+          // remaining case to handle is where it is a function.
+          continue;
+        }
+        aggr.get_active()->add(*(*input)->Slice(offset, length));
```

Comment:
@dominiklohmann We run into an assertion here for `read zeek-json | summarize x=distinct(id)`.
~~~
/Users/jannis/vast/libvast/include/vast/arrow_table_slice.hpp:292: assertion failed 'type.to_arrow_type()->id() == arr.type_id()'
0x707300010518ff0c: (caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>> vast::value_at<arrow::Array>(vast::type const&, arrow::Array const&, long long)+0x298)
0xc81100010518f5bc: (vast::view_trait<vast::type_to_data<vast::record_type>::type>::type vast::value_at<vast::record_type>(vast::record_type const&, vast::type_to_arrow_array_storage<vast::record_type>::type const&, long long)::record_view::at(unsigned long) const+0x210)
0x255c80010500e9c8: (vast::detail::container_view_iterator<std::__1::pair<std::__1::basic_string_view<char, std::__1::char_traits<char>>, caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>>>>::dereference() const+0x128)
0x246900010500e880: (decltype(auto) vast::detail::iterator_access::dereference<vast::detail::container_view_iterator<std::__1::pair<std::__1::basic_string_view<char, std::__1::char_traits<char>>, caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>>>>>(vast::detail::container_view_iterator<std::__1::pair<std::__1::basic_string_view<char, std::__1::char_traits<char>>, caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>>>> const&)+0x20)
0x542080010500e158: (vast::detail::iterator_facade<vast::detail::container_view_iterator<std::__1::pair<std::__1::basic_string_view<char, std::__1::char_traits<char>>, caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>>>>, std::__1::pair<std::__1::basic_string_view<char, std::__1::char_traits<char>>, caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>>>, std::__1::random_access_iterator_tag, std::__1::pair<std::__1::basic_string_view<char, std::__1::char_traits<char>>, caf::variant<caf::none_t, bool, long long, unsigned long long, double, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>, std::__1::chrono::time_point<std::__1::chrono::system_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>>>, std::__1::basic_string_view<char, std::__1::char_traits<char>>, vast::pattern_view, vast::ip, vast::subnet, unsigned char, vast::container_view_handle<vast::list_view_ptr>, vast::container_view_handle<vast::map_view_ptr>, vast::container_view_handle<vast::record_view_ptr>>>, long>::operator*() const+0x28)
0xf17380010500da48: (void vast::hash_append<vast::xxh3_64>(vast::xxh3_64&, vast::container_view_handle<vast::record_view_ptr>)+0x2bc)
0x2f0300010502c8b0: (unsigned long long vast::detail::hash_algorithm_proxy<vast::xxh3_64>::operator()<vast::container_view_handle<vast::record_view_ptr>>(vast::container_view_handle<vast::record_view_ptr> const&) const+0x1a8)
0x486780010502c658: (auto vast::hash<vast::xxh3_64, vast::container_view_handle<vast::record_view_ptr>>(vast::container_view_handle<vast::record_view_ptr> const&)+0x134)
0x3f738001050da0d4: (vast::plugins::distinct::(anonymous namespace)::heterogeneous_data_hash<vast::record_type>::operator()(vast::container_view_handle<vast::record_view_ptr>) const+0x1c)
...
~~~
The issue is that `id` is a record, but I'm not sure if this is rather something that requires a fix outside of `summarize`.

Reply:
This appears to happen when calling the hash function. Let me take a look.

Reply:
We found the problem, so this can be resolved now.